src/mono/netcore/System.Private.CoreLib/System.Private.CoreLib.csproj

```diff
@@ -275,7 +275,6 @@
       <Compile Include="$(BclSourcesRoot)\System\Threading\Thread.Mono.cs" />
       <Compile Include="$(BclSourcesRoot)\System\Threading\ThreadPool.Mono.cs" />
       <Compile Include="$(BclSourcesRoot)\System\Threading\LowLevelLock.cs" />
-      <Compile Include="$(BclSourcesRoot)\System\Threading\LowLevelSpinWaiter.cs" />
```

Comment:
Is `LowLevelLifoSemaphore.Unix.Mono.cs` still needed?

Reply:
Yea coreclr and mono currently have different ways of getting the LIFO wait release order. Coreclr's version creates a regular semaphore and the LifoSemaphore's wait issues a prioritized wait in the PAL such that the waiter is registered in stack order instead of the usual queue order, release works the same way as a dequeue, so prioritized waits would be released in LIFO order. If the waiter registration order could be similarly tweaked in mono then we can have a simpler runtime call to do that and share most of the implementations. On Windows coreclr uses IO completion port so it would continue have a different implementation but that may be shareable too for Windows.

Reply:
My preference would be to have same implementation. @lambdageek is this something we can easily tweak?

Reply:
@marek-safar if I follow the question is whether we can share most of the managed implementation which uses a normal managed semaphore and has a separate QCall that queues the waiting thread ahead of the others?  If yes, I think that will be a bit challenging for us.  At the lowest level we implement waiting on a semaphore by waiting on a pthread condition variable. pthread condition variables don't give us a way to affect the queue order (as far as I know).

Reply:
I see, how does `WaitHandle.WaitAll` with a bunch of semaphores work in mono?

It seems possible that mono's implementation may be simpler and more performant than coreclr's, just trying to understand differences and reasons for them if any.

Reply:
`WaitAll` is done by waiting on a global condvar. 

https://github.com/dotnet/runtime/blob/bc782a44d942c58799678f6bbac716a298928bce/src/mono/mono/metadata/w32handle.c#L40-L46

Reply:
I see, so when any wait object that is a component of a multi-wait is signaled, all threads waiting on it would be released and each thread would, for a wait-all, I imagine, acquire the mutex for each of the wait objects involved in deterministic order to determine if all of the waits are satisfied?

That might be an advantage in coreclr's PAL implementation (which is similar to Windows) where the signaling thread determines whether multi-waits of a waiter are satisfied before releasing it, thereby releasing a minimum number of threads.